cli/internal/k8s/gitops.go

```diff
@@ -0,0 +1,164 @@
+			// Load the file contents
+			data, err := ioutil.ReadFile(manifest)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+			// Split the k8s resources
+			items, err := kube.SplitYAML(data)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+			res = append(res, items...)
+
+		}
+	}
+
+	for i := range res {
+		annotations := res[i].GetAnnotations()
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[annotationGCMark] = syncSettings.getGCMark(kube.GetResourceKey(res[i]))
+		res[i].SetAnnotations(annotations)
+	}
+	return res, nil
+}
+
+func GitopsProcess(path string, revision string, namespace string) {
+
+	logContext := logrus.WithField("manifest", path)
+	syncSettings := settings{path}
+	restConfig := getRestConfig()
+
+	logContext.Info("Caching cluster state data")
+	clusterCache := cache.NewClusterCache(restConfig,
+		cache.SetPopulateResourceInfoHandler(func(un *unstructured.Unstructured, isRoot bool) (info interface{}, cacheManifest bool) {
+			// store gc mark of every resource
+			gcMark := un.GetAnnotations()[annotationGCMark]
+			info = &resourceInfo{gcMark: un.GetAnnotations()[annotationGCMark]}
+			// cache resources that has that mark to improve performance
+			cacheManifest = gcMark != ""
+			return
+		}),
+	)
+
+	gitOpsEngine := engine.NewEngine(restConfig, clusterCache)
+	ctx, done := context.WithCancel(context.Background())
+
+	_, err := gitOpsEngine.Run()
+	if err != nil {
+		logContext.Debug(err)
+		logContext.Fatal("Error syncing cluster state")
+	}
+
+	attempt := 0
+	for {
+		attempt++
+
+		if attempt > 20 {
+			logrus.Warn("Unable to complete manifest deployment")
+			break
+		}
+
+		logrus.Infof("Sync attempt %d of 20", attempt)
+
+		target, err := syncSettings.parseManifests()
+		if err != nil {
+			logrus.Debug(err)
+			logrus.Error(err, "Failed to parse target state")
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
+		result, err := gitOpsEngine.Sync(
+			ctx,
+			target,
+			func(r *cache.Resource) bool {
+				return r.Info.(*resourceInfo).gcMark == syncSettings.getGCMark(r.ResourceKey())
+			},
+			revision,
+			namespace,
+			sync.WithLogr(newLogrusLogger(logContext)),
+			sync.WithPrune(true),
+			sync.WithManifestValidation(true),
+			sync.WithNamespaceCreation(true, func(un *unstructured.Unstructured) bool {
+				return true
+			}),
+		)
+
+		if err != nil {
+			logrus.Debug(err)
+			logrus.Error(err, "Failed to synchronize cluster state")
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
+		for _, res := range result {
+			logrus.WithField("result", res.Message).Info(res.ResourceKey.String())
+		}
+
+		break
+	}
+
+	logContext.Info("Sync operations complete")
+	done()
+
+}
+
+// https://github.com/argoproj/argo-cd/blob/a21b0363e39e93982d280722a0eb86c449145766/util/log/logrus.go#L20
+func newLogrusLogger(fieldLogger logrus.FieldLogger) logr.Logger {
+	return adapter.NewLoggerWithFormatter(fieldLogger, func(val interface{}) string {
```

Comment:
I now see what you mean when you said the logging situation was weird..

To make sure I understand what this is doing. Is this essentially dynamically defining the `NewLogrusLogger` function argo-cd will call when they want to create a new logger?

Reply:
yeah basically they were using logrus in gitops engine but decided to make it more generic since klog was already used by the k8s libraries anyway.  So argocd made a little wrapper, it's not perfect, but it catches a lot of the noise at least in logrus.  The whole thing is super gross through and we may end up having to just capture it all from stdout.  🤮